From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: izLoki <izLoki011@gmail.com>
Date: Thu, 5 Sep 2024 13:46:27 -0300
Subject: [PATCH] item stack methods


diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index 9b8b958ce181b60eb5db89d4720380153152a2ac..d072800fcd0b157f02d40037a91370b0a083f92b 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -1591,4 +1591,10 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
         return random.nextInt(unbreaking + 1) > 0;
     }
     // Purpur end
+
+    //Prisma start
+    public void setShowTooltip(boolean showTooltip) {
+        this.craftDelegate.setShowTooltip(showTooltip);
+    }
+    //Prisma end
 }
